libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/PredictiveCacheController.kt

```diff
@@ -87,32 +68,51 @@ class PredictiveCacheController @JvmOverloads constructor(
     }
 
     /**
-     * Call when a new map instance is available. Only one map instance at a time is supported.
+     * Call when a new map instance is available.
      */
-    fun setMapInstance(map: MapboxMap) {
-        removeMapInstance()
+    @JvmOverloads
+    fun addMapInstance(map: MapboxMap, sourceIdsToCache: List<String> = emptyList()) {
         val tileStore = map.getResourceOptions().tileStore
         if (tileStore == null) {
             handleError("TileStore instance not configured for the Map.")
             return
         }
-        traverseMapSources(map) { tileVariant ->
-            PredictiveCache.createMapsController(tileStore, tileVariant)
+
+        mapListeners[map]?.let {
+            removeMapInstance(map)
+        }
+
+        traverseMapSources(map, sourceIdsToCache) { tileVariant ->
+            PredictiveCache.createMapsController(map, tileStore, tileVariant)
         }
+
+        val onStyleLoadedListener = OnStyleLoadedListener {
+            val currentMapSources = mutableListOf<String>()
+            traverseMapSources(map, sourceIdsToCache) { tileVariant ->
+                currentMapSources.add(tileVariant)
+            }
+            updateMapsControllers(
+                map,
+                currentMapSources,
+                PredictiveCache.currentMapsPredictiveCacheControllers(map),
+                tileStore
+            )
+        }
+
         map.addOnStyleLoadedListener(onStyleLoadedListener)
-        this.map = map
+        mapListeners[map] = onStyleLoadedListener
     }
 
     /**
      * Remove the map instance. Call this whenever the [MapView] is destroyed
      * to avoid leaking references or downloading unnecessary resources.
      */
-    fun removeMapInstance() {
-        map?.removeOnStyleLoadedListener(onStyleLoadedListener)
-        PredictiveCache.currentMapsPredictiveCacheControllers().forEach { tileVariant ->
-            PredictiveCache.removeMapsController(tileVariant)
+    fun removeMapInstance(map: MapboxMap) {
```

Comment:
Noticed that this is removing all map controllers. Should we rename the API accordingly?

Also noticed that we have a more granular API but we don't expose it publicly 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/1e4b0ea1033dbc3c6986f4052172dc3cf15f4992/libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/PredictiveCache.kt#L84

Should we? Is it a valid use case to remove map controllers separately? 🤔

Reply:
renamed to
`createMapControllers`
`removeMapControllers`

> Is it a valid use case to remove map controllers separately

don't think so, when we create them, we can specify what sources to cache, but when removing we don't it, just remove all of them.